bluepyentity/entity_definitions.py

```diff
@@ -0,0 +1,338 @@
+"""Datamodels with some of the functionality"""
+import datetime
+import inspect
+import pathlib
+import re
+import sys
+from abc import ABC, abstractmethod
+from typing import List, Literal
+
+import kgforge
+import pydantic
+
+MODULE = sys.modules[__name__]
+
+FILE_URI = "file://"
+RE_URL = re.compile("^http(s)?://")
+RE_NOT_FILE_URI = re.compile(f"^(?!{FILE_URI})")
+
+
+def _wrap_file_uri(path):
+    """Wraps path as file URI if not already wrapped"""
+    return RE_NOT_FILE_URI.sub(FILE_URI, path)
+
+
+def _is_url(item):
+    return isinstance(item, str) and RE_URL.match(item)
+
+
+def _fetch(id_, forge):
+    """Fetch a resource or a list of resources."""
+    if id_ is None:
+        return None
+    if _is_url(id_):
+        return forge.retrieve(id_, cross_bucket=True)
```

Comment:
The whole project-wise? Yeah, we should add this to utils.

This is part of older code. I think `_fetch` should no longer return `None` in any case since none of the ID-fields should be `None` at the time when they are fetched. Actually, I don't think it should check for `_is_url` either. It should either call `retrieve` on the given `id_` if it's a `str`, call itself recursively if it's a `list` and raise otherwise.

In the previous "version" of registration, you could either refer to field with an id or give a dictionary of a resource to be also registered:
```yml
<reference>: <id>
# OR
<reference>:
  type: AlsoRegisterMe
```
In this model, we currently only allow the former. We likely need to also allow the latter but in that case, registration would first be called for that subdict and the registered reference's `id` would then be added for the resource's `<reference>` field, so at this point, it would still  be either a `str`, `list`.